src/models/stores/sortable-documents.ts

```diff
@@ -0,0 +1,196 @@
+    const tagsWithDocs: Record<string, TagWithDocs> = {};
+    if (commentTags) {
+      for (const key of Object.keys(commentTags)) {
+        tagsWithDocs[key] = {
+          tagKey: key,
+          tagValue: commentTags[key],
+          docKeysFoundWithTag: []
+        };
+      }
+      tagsWithDocs[""] = { //this accounts for when user commented with tagPrompt (no tag selected)
+        tagKey: "",
+        tagValue: "Not Tagged",
+        docKeysFoundWithTag: []
+      };
+    }
+
+    this.tempTagDocumentMap.forEach((docKeysSet, tag) => {
+      const docKeysArray = Array.from(docKeysSet); // Convert the Set to an array
+      if (tagsWithDocs[tag]) {
+        tagsWithDocs[tag].docKeysFoundWithTag = docKeysArray;
+      }
+    });
+
+    const sortedDocsArr: SortedDocument[] = [];
+
+    Object.entries(tagsWithDocs).forEach((tagKeyAndValObj) => {
+      const tagWithDocs = tagKeyAndValObj[1] as TagWithDocs;
+      const sectionLabel = tagWithDocs.tagValue;
+      const docKeys = tagWithDocs.docKeysFoundWithTag;
+      const documents = this.documents.all.filter(doc => docKeys.includes(doc.key));
+      sortedDocsArr.push({
+        sectionLabel,
+        documents
+      });
+    });
+
+    return sortedDocsArr;
+  }
+
+  //Ideas with Scott:
+  //possibly could be an "auto-run"
+
+  updateTagDocumentMap () {
+    const db = this.db.firestore;
+    const filteredDocs = this.filteredDocsByType;
+
+    filteredDocs.forEach(doc => {
+      //Comments are stored in the Firestore document with `uid:teacher_documentKey`
+      //where uid:teacher is the teacher who made the comments on the document with documentKey
+      //docsRef may be a reference to multiple documents
+      //i.e one for student (who made document) plus any for teachers who commented
+      const docsRef = db.collection("documents").where("key", "==", doc.key);
+      docsRef.get().then(docsSnapshot => {
+        if (!docsSnapshot.empty){
+          docsSnapshot.forEach(docSnapshot =>{
+            const commentsRef = docSnapshot.ref.collection("comments");
+            commentsRef.get().then(commentsSnapshot => {
+              if (!commentsSnapshot.empty){
+                runInAction(() => {
+                  commentsSnapshot.forEach(commentDoc => {
+                    const commentData = commentDoc.data();
+                    if (commentData && commentData.tags) {
+                      commentData.tags.forEach((tag: string) => {
+                        let docKeysSet = this.tempTagDocumentMap.get(tag);
+                        if (!docKeysSet) {
+                          docKeysSet = new ObservableSet<string>();
+                          this.tempTagDocumentMap.set(tag, docKeysSet);
+                        }
+                        docKeysSet.add(doc.key); //only unique doc keys will be stored
+                      });
+                    }
+                  });
+                });
+              }
+            });
+          });
+        }
+      });
```

Comment:
If you change this to an async action the nesting of this code could be improved.
The `if(!docsSnapshot.empty)` and  ` if (!commentsSnapshot.empty)` are not needed. If they are empty then the loop below just won't run.
```suggestion
      const docsRef = db.collection("documents").where("key", "==", doc.key);
      const docsSnapshot = await docsRef.get();      
      docsSnapshot.forEach(async (docSnapshot) =>{
        const commentsRef = docSnapshot.ref.collection("comments");
        const commentsSnapshot = await commentsRef.get();
        runInAction(() => {
          commentsSnapshot.forEach(commentDoc => {
            const commentData = commentDoc.data();
            commentData?.tags?.forEach((tag: string) => {
              let docKeysSet = this.tempTagDocumentMap.get(tag);
              if (!docKeysSet) {
                docKeysSet = new ObservableSet<string>();
                this.tempTagDocumentMap.set(tag, docKeysSet);
              }
              docKeysSet.add(doc.key); //only unique doc keys will be stored
            });
          });
        });
      });
```